FLT/AutomorphicRepresentation/Example.lean

```diff
@@ -321,9 +321,48 @@ noncomputable abbrev zHatsub : AddSubgroup QHat :=
 noncomputable abbrev zsub : AddSubgroup QHat :=
   (Int.castRingHom QHat : ℤ →+ QHat).range
 
-lemma rat_meet_zHat : ratsub ⊓ zHatsub = zsub := sorry
+lemma rat_meet_zHat : ratsub ⊓ zHatsub = zsub := by
+  apply le_antisymm
+  · intro x ⟨⟨l, hl⟩, ⟨r, hr⟩⟩
+    simp at hl hr; rcases lowestTerms x with ⟨⟨N, z, hNz, hx⟩, unique⟩
+    have cop1 : IsCoprime l.den.toPNat' l.num := by
+      have : PNat.val l.den.toPNat' = l.den := by simp [l.den_pos]
+      simp [IsCoprime]; rw [this]
+      have hlp := (ZMod.isUnit_iff_coprime l.num.natAbs l.den).2 l.reduced
+      let m := l.num; have hm : l.num = m := by rfl
+      rcases m with m | m
+      · simp [hm] at *; exact hlp
+      · simp [hm] at *; rw [← neg_add, add_comm]; exact IsUnit.neg hlp
```

Comment:
If you add
```lean
lemma ZMod.isUnit_natAbs {z : ℤ} {N : ℕ} : IsUnit (z.natAbs : ZMod N) ↔ IsUnit (z : ZMod N) := by
  cases z.natAbs_eq with
  | inl h | inr h => rw [h]; simp [-Int.natCast_natAbs]
```
then you can rewrite this as
```suggestion
      simp_rw [IsCoprime, ZHat.intCast_val, ← ZMod.isUnit_natAbs, ZMod.isUnit_iff_coprime,
        PNat.toPNat'_coe l.den_pos]
      exact l.reduced
```
That's a lot of changes at one. I:

1. Replaced the `simp [IsCoprime]` by the output of `simp? [IsCoprime]`
2. Noted that the cases split didn't use anything specific about `l.num`, and tried to extract the general statement into the lemma above. That left me with `rw [ZMod.isUnit_natAbs] at hlp; exact hlp` to finish the proof.
3. Looked at `hlp` and noticed that I could move the `ZMod.isUnit_iff_coprime` application out: `have hlp := l.reduced; rw [← ZMod.isUnit_iff_coprime, ZMod.isUnit_natAbs] at hlp; exact hlp`
4. Rather than working on `hlp` until it matched the goal, I worked on the goal until it matched `hlp`: `rw [← ZMod.isUnit_natAbs, ZMod.isUnit_iff_coprime]; have hlp := l.reduced; exact hlp` 
5. Suspect that `this` should be provable with a single lemma, and found it by searching [loogle](https://loogle.lean-lang.org/?q=Nat.toPNat%27) for `Nat.toPNat'`

6. This got me to
```lean
      simp only [IsCoprime, ZHat.intCast_val]
      rw [PNat.toPNat'_coe l.den_pos]
      rw [← ZMod.isUnit_natAbs, ZMod.isUnit_iff_coprime]
      exact l.reduced
```
at which point I collapsed everything into a single `simp_rw`, though I had to move `PNat.toPNat'_coe l.den_pos` to the end for a reason I didn't investigate.